### DISCUSSION TITLE
Issues with Next.js Upgrade and Server-Side Components

### DISCUSSION THREAD

**Discussion by muhammad-umar-bhutta** [vercel/next.js#50730] [Help]
### Summary

Hello everyone! I hope you're all doing well. I'm currently facing some challenges after upgrading to the latest version of Next.js. I have been reading the official documentation, but there are a few solutions that I couldn't find. Previously, we used react-router, but now we have an app router in Next.js. The problem is that the new app router is missing the events.on.start and finish properties, which is causing the react-top-loading bar to not work in the newer version. If any of you have encountered a similar issue and have found a solution, I would greatly appreciate your guidance.

Additionally, Next.js has introduced a new feature called server-side components. While server-side components are working well, I'm facing an issue with defining metadata in the layout file and using it in specific pages like contact-us, sign-up, and sign-in. I need to use a client hook to access hooks in these pages, but when I attempt to use it, the metadata doesn't work due to the client-side component. As a workaround, I have to create a new component and move the metadata to it, which feels cumbersome.

Furthermore, I wanted to use server-side components, but I'm not very familiar with them yet. This lack of familiarity has led to another issue. I have a header component that is a client-side component, but it also contains a categories component that retrieves data from the server. I would like the categories component to render data on the client side, but when I use the client, the server-side component doesn't work and I encounter an error. While I understand that separating the components might be suggested, I'm also using the use effect hook in the categories component to perform specific actions on mouseover and leave, which makes it challenging to separate them.

I kindly request your assistance in resolving these issues. Your suggestions and guidance will be greatly appreciated in overcoming these challenges. Thank you in advance for your support.

**Comment by devjiwonchoi** [comment]
In the `App Router`, you can use the `loading.js` in order to display the page loader.

> A loading file can create instant loading states built on [Suspense](https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming).

Example:

```js
export default function Loading() {
  // Or a custom loading skeleton component
  return 'Loading...';
}
```

Since the `react-top-loading-bar` is a client-side feature, set `'use client'` on the top of the page in order to make the `loading.js` as a Client Component.

> By default, this file is a [Server Component](https://nextjs.org/docs/getting-started/react-essentials#server-components) - but can also be used as a Client Component through the "use client" directive.

Example:

```js
'use client'
import LoadingBar from 'react-top-loading-bar'

export default function Loading() {
  return (
   <LoadingBar />
  )
}
```

For more information, look [here](https://nextjs.org/docs/app/api-reference/file-conventions/loading).

**Reply by muhammad-umar-bhutta** [reply]
Yes, we can use a loading file for displaying progress, but in this case, I'm specifically referring to the react-top-loading progress bar. This progress bar shows the progress of the page, such as the start, middle, and end, indicated by percentages. I'm looking for a proper solution to incorporate this functionality. In the earlier version, its worked well cause of react-router events properties but in the app router, this option is not available anymore. So please assist me in this regard to fix this issue as soon as possible.

Additionally, I'm experiencing an issue with the 404 page. I've created a file called not-found.js, which is located in the root directory (inside the app directory) of the project. The redirection to the correct page is working, but it then goes into a loop, continuously refreshing the same page. I've even tried changing the name of the page to "notFound," but the error still persists. Moreover, I also checked the 404 error working fine if I removed the not-found.js but I need the desired page. Here's the code snippet for the 404 page:
import Link from 'next/link';

export default function notFound() {
  return (
    <>
      <div className="bg-gray-50 h-[70vh] flex justify-center items-center">
        <div className="container px-4 py-8">
          <div className="max-w-2xl mx-auto text-center">
            <h1 className="text-5xl font-bold font-heading">404</h1>
            <p className="text-2xl font-medium mb-8">Page Not Found</p>
            <p className="text-lg mb-4">The page you are looking for could not be found.</p>
            <p className="text-lg mb-4">Please check the URL or return to the <Link href="/"><span className="text-blue-600">homepage</span></Link>.</p>
          </div>
        </div>
      </div>
    </>
  );
}
Furthermore, I'd like to provide an overview of the project structure, which consists of the following directories and files:

Root Directory
layout.js
page.js
loading.js
not-found.js
components (contains various components)
slug-pages (includes slug pages)
marketing (contains about-us and contact-us folders, each with their respective pages)
I appreciate any assistance you can provide in resolving these issues. Thank you for your support.
